content/blog/2025/05/esp-idf-mosquitto-port/index.md

```diff
@@ -0,0 +1,156 @@
+- Create resilient IoT networks that can operate across geographical locations
+- Enable MQTT communication between networks separated by NATs or firewalls
+
+
+## Technical Specifications
+- **Memory Footprint**:
+  - Program memory: ~60 kB
+  - Stack size: Minimum 5 kB (recommended)
+  - Heap usage: ~2 kB on startup, ~4 kB per connected client
+- **Connectivity**: TCP or TLS transport
+- **API**: Simple C API with broker configuration structure
+- **Dependencies**:
+  - ESP-IDF components: newlib, esp_timer, esp_tls (for secure TLS connections)
+  - Managed components: sock_utils (for porting)
+
+
+## Implementation Details
+The broker is implemented as a port of the original Mosquitto codebase. It follows a single-threaded, event-driven architecture that efficiently handles multiple connections through socket multiplexing. The core of this implementation consists of several key components:
+
+### Socket Multiplexing
+The broker uses an abstracted multiplexing layer that can utilize either poll-based or epoll-based implementations:
+- **Mux Layer**: An abstraction layer (`mux.c`) that provides common interface functions for socket multiplexing
+- **Poll Implementation**: Uses standard poll() mechanisms for handling multiple file descriptors
+- **Epoll Implementation**: An optimized event notification mechanism (supported on some systems)
+
+The ESP32 implementation uses the poll-based mechanism which is well-suited for the number of connections the ESP32 can realistically handle.
+
+### Event Loop
+The main broker loop (`loop.c`) performs the following operations:
+1. Processes incoming client connections
+2. Handles socket read/write events
+3. Manages client keepalive tracking
+4. Processes message queues and subscriptions
+5. Handles disconnections and cleanup
+
+### ESP-TLS Integration
+The port includes a custom implementation (`net__esp_tls.c`) that integrates ESP-TLS for secure connections:
+- Handles TLS context initialization
+- Manages secure socket connections
+- Processes TLS handshakes and secure data transfer
+- Maintains a fixed array of TLS contexts for connected clients
+
+### Simplified message handling
+```mermaid
+graph TD
+    A[Client Connection Request] --> B[Socket Accept]
+    B --> C{TLS Enabled?}
+    C -->|Yes| D[ESP-TLS Context Creation]
+    C -->|No| E[Plain TCP Connection]
+    D --> F[TLS Handshake]
+    F --> G[Connection Ready]
+    E --> G
+    G --> H[Socket Multiplexing]
+
+    subgraph "Event Loop"
+        H --> I[Poll for Events]
+        I --> J[Process Socket Events]
+        J --> K[Handle Client Messages]
+        K --> L[Route Messages to Subscribers]
+        L --> I
+    end
+
+    M[Client Disconnect] --> N[Socket Cleanup]
+    N --> O[Release Resources]
+```
+
+## Getting Started
+To use the ESP32 Mosquitto Port in your projects, you'll need to initialize the broker with appropriate configuration settings and provide sufficient stack space for the running task.
```

Comment:
I'm missing the step of adding the component from the component manager.

Reply:
Have added, thanks!